crates/flux-middle/src/rty/subst.rs

```diff
@@ -396,3 +396,23 @@ impl TypeFolder for SortSubst<'_> {
         }
     }
 }
+
+pub(crate) struct GenericSortSubst<'a> {
+    args: &'a [Sort],
+}
+
+impl<'a> GenericSortSubst<'a> {
+    pub(crate) fn new(args: &'a [Sort]) -> Self {
+        Self { args }
+    }
+}
+
+impl TypeFolder for GenericSortSubst<'_> {
+    fn fold_sort(&mut self, sort: &Sort) -> Sort {
+        if let Sort::Param(var) = sort {
+            self.args[var.index as usize].clone()
+        } else {
+            sort.super_fold_with(self)
+        }
+    }
+}
```

Comment:
This should go in `GenericsSubstFolder` so we only have one substitution for generics. We'll need to modify the logic in `GenericsSubstFolder::sort_for_param` to call `peel_out_sort`. Note that it's ok to panic if `peel_out_sort` returns `None` because we assume we only substitute on well-formed terms and that substitution preserves well-formedness. If substitution fails then the problem is somewhere else.

As a side note, the way I understand `Sort::Param` is as follows: if we have `|x: sortof(Self)| ...` and we substitute `Self` with `i32` we get `|x: sortof(i32)| ...` which then normalizes to `|x: int| ...`. Instead of having `x: sortof(i32)` as an intermediate step, we normalize it immediately as we apply the substitution, this way we don't need to represent `sortof` for arbitrary types, only for type parameters.

Reply:
Done, I think, but see the note in `check_alias_pred_app`